---
title: "Single-cell Analysis of Host-Microbiome Interactions"
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/mire)](https://CRAN.R-project.org/package=mire)
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![](https://cranlogs.r-pkg.org/badges/mire)](https://cran.r-project.org/package=mire)
<!-- badges: end -->

An integrated framework for high-resolution microbiome analysis in single-cell
sequencing. Leveraging Kraken2 for initial taxonomic classification, the package
performs barcode-aware filtering and reconstructs read-level information by
integrating cell barcodes and UMIs.  It quantifies k-mers and aggregates
taxon-specific signals across multiple taxonomic ranks. This enables accurate
reconstruction and quantitative profiling of microbial communities at
single-cell resolution.

## Installation
You can install `mire` from `CRAN` using:
```{r, eval=FALSE}
# install.packages("pak")
pak::pak("mire")
```

Alternatively, install the development version from
[r-universe](https://yunuuuu.r-universe.dev/mire) with:

```{r, eval=FALSE}
pak::repo_add("https://yunuuuu.r-universe.dev")
pak::pak("mire")
```

or from [GitHub](https://github.com/Yunuuuu/mire) with:
```{r, eval=FALSE}
pak::pak("Yunuuuu/mire")
```

You must also install [kraken2](https://github.com/DerrickWood/kraken2/wiki/Manual).

## sessionInfo
```{r}
sessionInfo()
```
